Replace base64.*string() functions to fix py3.9 support

mgorny opened the pull request:
Replace base64.decodestring() and .encodestring() functions as they
were deprecated since Python 3.1 in favor of (equivalent) .decodebytes()
and .encodebytes(), and were eventually removed in Python 3.9.

While at it, replace most of their uses with base64.b64encode()
and .b64decode() that are preferable to the former wrt ticket #6446,
and they do not introduce line breaks that the twisted code usually
discarded.

Use .decodebytes() and .encodebytes() in DirDBM as it seems to rely
on the exact presence of newlines, and changing that would break
backwards compatibility.

Fixes: ticket:6446
Fixes: ticket:9831

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/9831
* [x] The changes pass minimal style checks (see: https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted )
  * Note: the lint test failed for me but on existing lines unrelated to my change
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

mgorny commented:
I don't think these CI failures are related to my changes.

wiml reviewed (approved):
Also nice that this cleans up a bunch of `strip` calls that aren't needed with the newer api. Looks good to me!

wiml commented:
A question for @hawkowl or someone more familiar with towncrier than I am: is it a problem that the same bug number is mentioned in multiple newsfragments in subdirectories? Should they be moved up to the top level?

hawkowl commented:
(the CI failures are related to Azure Pipelines giving Unicode but the Python being configured to output cp1252 by default on Windows :/ -- been trying to debug it and come up with a solution)

Generally, unless there's a specific feature change (e.g. multiple external APIs in different sections), I collapse them into a single one in the main folder. I kind of want to get rid of the subpackages newsfragments, since it doesn't feel terribly useful now? Not sure.

glyph commented:
> (the CI failures are related to Azure Pipelines giving Unicode but the Python being configured to output cp1252 by default on Windows :/ -- been trying to debug it and come up with a solution)

What do you mean by "giving" unicode?  Giving unicode to what?  It seems like there's one failure:

```pytb
[ERROR]
Traceback (most recent call last):
  File "d:\a\1\s\build\alldeps-withcov-windows\lib\site-packages\twisted\test\test_process.py", line 2412, in gotEnvironment
    environ[self.goodKey.encode('ascii')],
builtins.KeyError: b'UNICODE'

twisted.test.test_process.Win32UnicodeEnvironmentTests.test_encodableUnicodeEnvironment
```

`environ` is always bytes-keys, right?

hawkowl commented:
> environ is always bytes-keys, right?

It fails because it raises an encoding error which gets printed to stdout because sys.stdout is cp1252 🙃

rodrigc commented:
I think that `Win32UnicodeEnvironmentTests.test_encodableUnicodeEnvironment`
is bogus and should be deleted.  By default `os.environ` is Unicode on all platforms in Python 3. 

There is not guarantee that all the environment variables in the environment will be ASCII.

Specifically what seems to be triggering this bug is that  **Michał Górny**'s name has non-ASCII characters in it.  Someone, when Azure Pipelines picks up this PR to build, something from Michał's name enters the environment, and triggers this one test failure on Windows.

I worked around this problem in: https://github.com/twisted/twisted/pull/1276
by amending the commit on Michał 's branch to have my name.

rodrigc commented:
> ```
> 
> `environ` is always bytes-keys, right?

No, `os.environ` is always Unicode in Python 3.

mgorny commented:
FTR, you can change my name to 'Michal Gorny'.

rodrigc committed:
Replace base64.*string() functions to fix py3.9 support

Replace base64.decodestring() and .encodestring() functions as they
were deprecated since Python 3.1 in favor of (equivalent) .decodebytes()
and .encodebytes(), and were eventually removed in Python 3.9.

While at it, replace most of their uses with base64.b64encode()
and .b64decode() that are preferable to the former wrt ticket #6446,
and they do not introduce line breaks that the twisted code usually
discarded.

Use .decodebytes() and .encodebytes() in DirDBM as it seems to rely
on the exact presence of newlines, and changing that would break
backwards compatibility.

Submitted by: Michal Gorny <mgorny@gentoo.org>

rodrigc committed:
Pass the correct os.environ dictionary to GetEnvironmentDictionary()

Also, remove an extraneous comment about Python 2, since the minimum
Python version that Twisted supports is Python 3.5

rodrigc committed:
Add more debugging in case of bad arg or env variables which can't be
decoded

rodrigc commented:
Fixed in https://github.com/twisted/twisted/pull/1304
